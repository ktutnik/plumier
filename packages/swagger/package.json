{
  "name": "@plumier/swagger",
  "version": "1.0.0-beta.9",
  "description": "Swagger and Open API 3 module for Plumier",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "compile": "tsc -p tsconfig.build.json"
  },
  "author": "Ketut Sandiarsa",
  "license": "MIT",
  "dependencies": {
    "@plumier/core": "1.0.0-beta.9",
    "openapi3-ts": "^1.3.0"
  },
  "bugs": {
    "url": "https://github.com/plumier/plumier/issues"
  },
  "repository": {
    "url": "https://github.com/plumier/plumier"
  },
  "homepage": "https://plumierjs.com",
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@types/dotenv": "^6.1.1",
    "dotenv": "^8.1.0"
  }
}
