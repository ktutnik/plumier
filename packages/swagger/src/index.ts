import {
    ActionResult,
    DefaultFacility,
    HttpMethod,
    Invocation,
    Middleware,
    PlumierApplication,
    response,
    RouteInfo,
} from "@plumier/core"
import { ServeStaticMiddleware } from "@plumier/serve-static"
import { OpenApiBuilder, OperationObject, PathItemObject, PathObject, ResponseObject } from "openapi3-ts"
import dist from "swagger-ui-dist"
import { join } from 'path'

// --------------------------------------------------------------------- //
// ------------------------------ HELPERS ------------------------------ //
// --------------------------------------------------------------------- //

interface RouteGroup { [url: string]: RouteInfo[] }

function groupRoutes(routes: RouteInfo[]): RouteGroup {
    return routes.reduce((prev, cur) => {
        prev[cur.url] = (prev[cur.url] || []).concat(cur)
        return prev
    }, {} as RouteGroup)
}

function transformUrl(url: string) {
    return url.replace(/(:(\w*\d*)(\/|-{0,1}))/g, (a, b, par, slash) => `{${par}}${slash}`)
}

// --------------------------------------------------------------------- //
// ----------------------------- TRANSFORM ----------------------------- //
// --------------------------------------------------------------------- //

function transformPaths(group: RouteGroup) {
    return Object.keys(group)
        .map(x => transformPath(x, group[x]))
        .reduce((result, [path, item]) => {
            result[path] = item
            return result
        }, {} as PathObject)
}

function transformPath(path: string, route: RouteInfo[]): [string, PathItemObject] {
    const item = route.map(x => transformOperation(x))
        .reduce((result, [method, opr]) => {
            result[method] = opr
            return result
        }, {} as PathItemObject)
    return [transformUrl(path), item]
}

function transformOperation(route: RouteInfo): [HttpMethod, OperationObject] {
    const operation: OperationObject = {
        responses: transformResponses(route)
    }
    return [route.method, operation]
}

function transformResponses(route: RouteInfo): { [status: string]: ResponseObject } {
    return {
        "200": { description: "", content: {} }
    }
}

function transform(routes: RouteInfo[]) {
    const group = groupRoutes(routes)
    const paths = transformPaths(group)
    return OpenApiBuilder.create({
        openapi: "3.0.0",
        info: { title: "title", version: "1.0.0" },
        paths
    }).getSpec()
}

// --------------------------------------------------------------------- //
// ----------------------------- MIDDLEWARE ---------------------------- //
// --------------------------------------------------------------------- //

export interface OpenApiFacilityConfiguration { endpoint: string }

class OpenApiMiddleware implements Middleware {
    constructor(private spec: any, private opt: OpenApiFacilityConfiguration) { }
    async execute(invocation: Readonly<Invocation>): Promise<ActionResult> {
        const uiPath = this.opt.endpoint.toLowerCase() + "/index"
        if (invocation.context.path.toLowerCase() === "/swagger.json")
            return response.json(this.spec)
        if (invocation.context.path.toLowerCase() === this.opt.endpoint.toLowerCase())
            return response.redirect(uiPath)
        if (invocation.context.path.toLowerCase() === uiPath)
            return response.file(join(dist.getAbsoluteFSPath(), "index.html"))
        else
            return invocation.proceed()
    }
}

export class OpenApiFacility extends DefaultFacility {
    opt: OpenApiFacilityConfiguration
    constructor(opt?: OpenApiFacilityConfiguration) {
        super()
        this.opt = { endpoint: "/swagger", ...opt }
        if(this.opt.endpoint.toLocaleLowerCase().endsWith("/index")) 
            this.opt.endpoint = this.opt.endpoint.substring(0, this.opt.endpoint.length - 6)
    }

    async initialize(app: Readonly<PlumierApplication>, routes: RouteInfo[]): Promise<void> {
        const path = dist.getAbsoluteFSPath()
        const spec = transform(routes)
        app.use(new OpenApiMiddleware(spec, this.opt))
        app.use(new ServeStaticMiddleware({ root: path, rootPath: this.opt.endpoint }))
    }
}