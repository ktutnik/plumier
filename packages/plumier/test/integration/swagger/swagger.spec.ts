import { OpenApiFacility, OpenApiFacilityConfiguration } from "@plumier/swagger"
import Plumier, { Class, WebApiFacility, route } from "plumier"
import supertest = require("supertest")

export function appStub(controller: Class | Class[] | string, opt?: OpenApiFacilityConfiguration) {
    return new Plumier()
        .set({ mode: "production" })
        .set(new WebApiFacility({ controller }))
        .set(new OpenApiFacility(opt))
        .initialize()
}

describe("Open API Paths", () => {
    it("Should provide path on controller convention", async () => {
        class AnimalController {
            index() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should provide correct path template", async () => {
        class AnimalController {
            @route.get(":id")
            index(id: string) { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should provide correct path template in nested route", async () => {
        @route.root("/users/:userId/animals")
        class AnimalController {
            @route.get(":id")
            index(id: string) { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should response http method on controller decorated with @route.get()", async () => {
        class AnimalController {
            @route.get("")
            get() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should response http method on controller decorated with @route.post()", async () => {
        class AnimalController {
            @route.post("")
            post() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should response http method on controller decorated with @route.put()", async () => {
        class AnimalController {
            @route.put("")
            put() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should response http method on controller decorated with @route.patch()", async () => {
        class AnimalController {
            @route.patch("")
            patch() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should response http method on controller decorated with @route.delete()", async () => {
        class AnimalController {
            @route.delete("")
            delete() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should response http method on controller decorated with @route.head()", async () => {
        class AnimalController {
            @route.head("")
            head() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should response http method on controller decorated with @route.options()", async () => {
        class AnimalController {
            @route.options("")
            options() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })

    it("Should response http method on controller decorated with @route.trace()", async () => {
        class AnimalController {
            @route.trace("")
            trace() { }
        }
        const app = await appStub(AnimalController)
        const { body } = await supertest(app.callback())
            .get("/swagger.json")
            .expect(200)
        expect(body.paths).toMatchSnapshot()
    })
})
