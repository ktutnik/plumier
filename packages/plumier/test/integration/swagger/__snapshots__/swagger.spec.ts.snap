// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Open API Paths Should provide correct path template 1`] = `
Object {
  "/animal/{id}": Object {
    "get": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should provide correct path template in nested route 1`] = `
Object {
  "/users/{userid}/animals/{id}": Object {
    "get": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should provide path on controller convention 1`] = `
Object {
  "/animal/index": Object {
    "get": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should response http method on controller decorated with @route.delete() 1`] = `
Object {
  "/animal": Object {
    "delete": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should response http method on controller decorated with @route.get() 1`] = `
Object {
  "/animal": Object {
    "get": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should response http method on controller decorated with @route.head() 1`] = `
Object {
  "/animal": Object {
    "head": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should response http method on controller decorated with @route.options() 1`] = `
Object {
  "/animal": Object {
    "options": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should response http method on controller decorated with @route.patch() 1`] = `
Object {
  "/animal": Object {
    "patch": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should response http method on controller decorated with @route.post() 1`] = `
Object {
  "/animal": Object {
    "post": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should response http method on controller decorated with @route.put() 1`] = `
Object {
  "/animal": Object {
    "put": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;

exports[`Open API Paths Should response http method on controller decorated with @route.trace() 1`] = `
Object {
  "/animal": Object {
    "trace": Object {
      "responses": Object {
        "200": Object {
          "content": Object {},
          "description": "",
        },
      },
    },
  },
}
`;
